src/storage/src/hummock/store/state_store.rs

```diff
@@ -12,57 +12,303 @@
+        self.read_version.write().update(info)
+    }
+
+    pub async fn get_inner<'a>(
+        &'a self,
+        key: &'a [u8],
+        epoch: u64,
+        read_options: ReadOptions,
+    ) -> StorageResult<Option<Bytes>> {
+        use parking_lot::RwLockReadGuard;
+        let compaction_group_id = match read_options.table_id.as_ref() {
+            None => None,
+            Some(table_id) => Some(self.get_compaction_group_id(*table_id).await?),
+        };
+
+        let key_range = (Bound::Included(key.to_vec()), Bound::Included(key.to_vec()));
+
+        let (staging_imm, staging_sst, committed_version) = {
+            let read_version = self.read_version.read();
+
+            let (staging_imm_iter, staging_sst_iter) =
+                read_version
+                    .staging()
+                    .prune_overlap(epoch, compaction_group_id, &key_range);
+
+            let staging_imm = staging_imm_iter
+                .cloned()
+                .collect::<Vec<ImmutableMemtable>>();
+
+            let staging_sst = staging_sst_iter.cloned().collect::<Vec<_>>();
+            let committed_version = read_version.committed().clone();
+
+            RwLockReadGuard::unlock_fair(read_version);
+            (staging_imm, staging_sst, committed_version)
+        };
+
+        let mut table_counts = 0;
+        let internal_key = key_with_epoch(key.to_vec(), epoch);
+        let mut local_stats = StoreLocalStatistic::default();
+
+        // 1. read staging data
+        // 2. use rev to ensure read the newest data first
+        for imm in staging_imm {
+            if imm.epoch() != epoch {
+                continue;
+            }
+
+            if !range_overlap(&(key..=key), imm.start_user_key(), imm.end_user_key()) {
+                continue;
+            }
+
+            if let Some(data) = get_from_batch(&imm, key, &mut local_stats) {
+                return Ok(data.into_user_value());
+            }
+        }
+
+        for local_sst in staging_sst {
+            let table = self
+                .sstable_store
+                .sstable(&local_sst, &mut local_stats)
+                .await?;
+
+            if let Some(data) = get_from_table(
+                self.sstable_store.clone(),
+                table,
+                &internal_key,
+                read_options.check_bloom_filter,
+                &mut local_stats,
+            )
+            .await?
+            {
+                return Ok(data.into_user_value());
+            }
+        }
+
+        for level in committed_version.levels(compaction_group_id) {
+            if level.table_infos.is_empty() {
+                continue;
+            }
+            match level.level_type() {
+                LevelType::Overlapping | LevelType::Unspecified => {
+                    let table_infos = prune_ssts(level.table_infos.iter(), &(key..=key));
+                    for table_info in table_infos {
+                        let table = self
```

Comment:
Shall we extract the following logic out to a new method or a closure? They are all only searching a key in a SST.

```
let table = self
      .sstable_store
      .sstable(table_info, &mut local_stats)
      .await?;
  if let Some(v) = get_from_table(
      self.sstable_store.clone(),
      table,
      &internal_key,
      read_options.check_bloom_filter,
      &mut local_stats,
  )
  .await?
```